mulog-prometheus/src/com/brunobonacci/mulog/publishers/prometheus/registry.clj

```diff
@@ -0,0 +1,76 @@
+(ns com.brunobonacci.mulog.publishers.prometheus.registry
+  (:require [com.brunobonacci.mulog.publishers.prometheus.collector :as col])
+  (:import [io.prometheus.client CollectorRegistry]
+           [io.prometheus.client.exporter.common TextFormat]))
+
+(defonce ^:private names-to-collectors (-> CollectorRegistry
+                                         (.getDeclaredField "namesToCollectors")
+                                         (doto (.setAccessible true))))
+
+(defonce ^:private names-collectors-lock (-> CollectorRegistry
+                                           (.getDeclaredField "namesCollectorsLock")
+                                           (doto (.setAccessible true))))
+
+
+
+(defprotocol Registry
+  "This protocol is used to extend the `CollectorRegistry`.
+  This is done to ultimately ensure thread safety and to dynamically register collections.
+  The prometheus java client currently only lets you register new collections, existing
+  collections will throw an `IllegalArgumentException`"
+  (get-nc-map
+    [t]
+    "Retrieve the `namesToCollectors` map")
+  (get-nc-lock
+    [t]
+    "Retrieve the `nameCollectorsLock` lock Object")
+  (register-dynamically
+    [t metric]
+    "This will try and register a new collection if it doesn't or return an existing collection
+    by doing the following:
+    - syncronize a lock on `nameCollectorsLock`
+    - get collection from `namesToCollectors` using `:metric/full-name`
+    - if collection exists return
+    - else register new collection and return"))
+
+(defprotocol ReadRegistry
+  "This protocol is used to extract the metric information from the registry."
+  (get-registry
+    [t]
+    "Returns the `^CollectorRegistry t`.")
+  (write-out
+    [t out]
+    "Writes the `^CollectorRegistry t` to `^java.io.Writer out`.")
+  (write-text
```

Comment:
since you return a string it should be `write-str` like the function `prn` and `prn-str`.

Reply:
Yes thanks based on the naming convention links you have provided updated accordingly.